# This script contains:
1. Import libraries and customer data
2. Finding basic data shape and printing head rows
3. Renaming columns
4. Basic data descriptions
5. Checking for duplicates
6. Finding mixed data
7. Checking for missing values
8. Combining customer_data with previously prepared data

# 1. Import libraries and customer data

In [97]:
# Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [99]:
# Defining Path
path= r'/Users/navneetjha/Desktop/Career foundry Docs/exercises/Data Immersion /Achievement 4/INSTACART BASKET ANALYSIS JULY14-2024'

In [101]:
path

'/Users/navneetjha/Desktop/Career foundry Docs/exercises/Data Immersion /Achievement 4/INSTACART BASKET ANALYSIS JULY14-2024'

In [103]:
# Import Customer data
customer_data=pd.read_csv(os.path.join(path,'02 DATA','ORIGINAL','Raw','customers.csv'), index_col=False)

# 2. Finding basic data shape and printing head rows

In [106]:
# Finding shape
customer_data.shape

(206209, 10)

In [108]:
# Printing head rows
customer_data.head(15)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559

# 3. Renaming columns

In [111]:
# Renaming columns
customer_data=customer_data.rename(columns={'Surnam':'surname','First Name':'first name','STATE':'state','Age':'age','Gender':'gender','n_dependants':'number_of_dependents','fam_status':'marital_status'}, inplace=False)

In [113]:
customer_data.head()

user_id first name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

# 4. Basic data descriptions

In [19]:
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   first name            194950 non-null  object
 2   surname               206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   number_of_dependents  206209 non-null  int64 
 8   marital_status        206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


This data shows that there are rows that are missing first name

In [22]:
# Checking basic statistical values
customer_data.describe()

user_id            age  number_of_dependents         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

 The minimum age is 18 and maximum is 81.
 The minimum number of dependents is 0(could be possible in case of single) and maximum is 3.
 Mean income is 94632 which is slightly greater than the 50% quartile. The reason could be the high maximum value of income which is 593901. 

# 5. Checking for duplicates


In [26]:
#Checking Duplicates
customer_data_dups = customer_data[customer_data.duplicated()]

In [28]:
customer_data_dups

Empty DataFrame
Columns: [user_id, first name, surname, gender, state, age, date_joined, number_of_dependents, marital_status, income]
Index: []

There are no duplicates in the dataset

# 6.Finding mixed data

In [32]:
for col in customer_data.columns.tolist():
  weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer_data[weird]) > 0:
    print (col)

/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].ap

first name


/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].applymap(type) != customer_data[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/38/3t8_0dvn1h950p6h5yqb56rm0000gn/T/ipykernel_803/2918866651.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer_data[[col]].ap

The result shows that first name column has mixed data

In [35]:
# Fixing mixed data
customer_data['first name']=customer_data['first name'].astype('str')

In [37]:
# Checking dataype for 'First Name' column
customer_data['first name'].dtype

dtype('O')

In [39]:
# Changing the data type of "user_id" as we don't need to conduct any statistical analysis of that column
customer_data['user_id']=customer_data['user_id'].astype('str')

In [41]:
# checking data type for user_id
customer_data['user_id'].dtype

dtype('O')

O symbolises object. Thus all data are in required format.

# 7. Checking for missing values

In [45]:
customer_data.isnull().sum()

user_id                 0
first name              0
surname                 0
gender                  0
state                   0
age                     0
date_joined             0
number_of_dependents    0
marital_status          0
income                  0
dtype: int64

There are no missing data as well

# 8. Combining customer_data with previously prepared data

In [49]:
# Importing Previous combined data
ords_prods_merge=pd.read_pickle(os.path.join(path,'02 DATA', 'PREPARED DATA','Orders_Products_Merged_Grouped_AggregatedDATA.pkl'))

In [50]:
ords_prods_merge.shape

(32434212, 22)

In [51]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  ...  \
0   3139998      138            28          6                 11  ...   
1   1977647      138            30          6                 17  ...   
2    389851      709             2          0                 21  ...   
3    652770      764             1          3                 13  ...   
4   1813452      764             3          4                 17  ...   

   reordered  _merge  busiest_days busiest_period_of_day max_order  \
0          0    both  Regular Days           Most orders        32   
1          1    both  Regular Days        Average Orders        32   
2          0    both  Busiest Days        Average Orders         5   
3          0    both  Slowest Days           Most orders         3   
4          1    both  Slowest Days        Average Orders         3   

       loyalty_flag  mean_product_price spending_flag  \
0  Regular customer            6.935811   Low spender   
1  Regular customer            6.935811   Low spender   
2      New customer            7.930208   Low spender   
3      New customer            4.972414   Low spender   
4      New customer            4.972414   Low spender   

   median_days_between_orders    order_frequency  
0                         8.0  Frequent customer  
1                         8.0  Frequent customer  
2                         8.0  Frequent customer  
3                         9.0  Frequent customer  
4                         9.0  Frequent customer  

[5 rows x 22 columns]

In [52]:
customer_data.head()

user_id first name   surname  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

The common column in both dataset is "user_id". So we have to perform the join on "user_id".

In [54]:
# Checking data type of linking column in both datasets
customer_data['user_id'].dtype

dtype('O')

In [60]:
ords_prods_merge['user_id'].dtype

dtype('int64')

dtype of "user_id" in "ords_prods_merge" dataframe is "int64" and in "customer_data" is "object". Therefore, we need to first change the data type of "user_id" in "ords_prods_merge" dataframe.

In [63]:
#Changing data type of "user_id" in "ords_prods_merge" dataframe.
ords_prods_merge['user_id']=ords_prods_merge['user_id'].astype('str')

In [65]:
# Checking data type 
ords_prods_merge['user_id'].dtype

dtype('O')

The data type is uniform in both data sets.

In [68]:
# Merging the two dataframes
df_merged=ords_prods_merge.merge(customer_data,on='user_id',indicator=False)

In [70]:
# Checking shape of new merged file
df_merged.shape

(32434212, 31)

In [72]:
# Checking number of columns in newly merged file
df_merged.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day',
       'days_since_prior_order', 'add_to_cart_order', 'reordered', '_merge',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'mean_product_price', 'spending_flag', 'median_days_between_orders',
       'order_frequency', 'first name', 'surname', 'gender', 'state', 'age',
       'date_joined', 'number_of_dependents', 'marital_status', 'income'],
      dtype='object')

In [74]:
df_merged.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id user_id  order_number  order_dow  order_hour_of_day  ...  \
0   3139998     138            28          6                 11  ...   
1   1977647     138            30          6                 17  ...   
2    389851     709             2          0                 21  ...   
3    652770     764             1          3                 13  ...   
4   1813452     764             3          4                 17  ...   

     order_frequency  first name  surname  gender      state age  date_joined  \
0  Frequent customer     Charles      Cox    Male  Minnesota  81     8/1/2019   
1  Frequent customer     Charles      Cox    Male  Minnesota  81     8/1/2019   
2  Frequent customer     Deborah    Glass  Female    Vermont  66    6/16/2018   
3  Frequent customer     Heather    Myers  Female  Wisconsin  40     2/9/2020   
4  Frequent customer     Heather    Myers  Female  Wisconsin  40     2/9/2020   

  number_of_dependents  marital_status  income  
0                    1         married   49620  
1                    1         married   49620  
2                    2         married  158302  
3                    3         married   31308  
4                    3         married   31308  

[5 rows x 31 columns]

In [76]:
# Inner Join count
df_merged['_merge'].value_counts()

_merge
both          32434212
left_only            0
right_only           0
Name: count, dtype: int64

All joins are inner joins

In [79]:
df_merged.dtypes

product_id                       int64
product_name                    object
aisle_id                         int64
department_id                    int64
prices                         float64
order_id                         int64
user_id                         object
order_number                     int64
order_dow                        int64
order_hour_of_day                int64
days_since_prior_order         float64
add_to_cart_order                int64
reordered                        int64
_merge                        category
busiest_days                    object
busiest_period_of_day           object
max_order                        int64
loyalty_flag                    object
mean_product_price             float64
spending_flag                   object
median_days_between_orders     float64
order_frequency                 object
first name                      object
surname                         object
gender                          object
state                    

In [81]:
# Exporting the merged file in pickle format
df_merged.to_pickle(os.path.join(path,'02 DATA','PREPARED DATA','customer_order_product_allclean_mergedfile.pkl'))